### PR TITLE
Adding new state to leader election 

### PR DESCRIPTION
**Pull request by cryptoe**
Currently, in the Druid Leader selector, we have 2 states only. Either the process is a leader or not. 
But we need a 3rd state which represents an initialized leader on which external callers should gate on.

According to the zookeeper, once we receive the callback inside `LeaderLatchListener#isLeader()` that candidate is the leader. In Druid, we can take some time in running the `DruidLeaderSelector#listener` since there are a lot of DB operations done in both the overlord and the coordinator which can take time. Also for all "endpoints", a leader is usable only when it's initialized. 

 
 Another race condition was how `CuratorDruidLeaderSelector` was updating the boolean variable `leader`. 
 As `DruidLeaderSelector#listener#becomeLeader()` can take some time to run, zk can choose to change leadership behind the scenes. So now we can have two processes thinking they are the leaders. 
 
 Changed `CuratorDruidLeaderSelector#isLeader()` to gate on the "latchStatus" rather than the local boolean `leader` variable. 
 
As the `leader` variable is updated in a single-threaded executor service the following situation can very well happen. 
Lets say we have two overlords foo, bar. 

```
 t0: foo is chosen the leader by ZK,  print "Grayskull"----------------------------------------t10: initialized 
 ----t1: notLeader listener is called but is waiting cause exec is single threaded -------------------t11: bar isNotLeader()-> leader =false
 -------t2: bar is leader, print "Grayskull"-----------t5: initialized 
```
Between t5-t11, both foo,bar think they are the leaders. 

##### Key changed/added classes in this PR
 * `DruidLeaderSelector`
 * `CuratorLeaderSelector`
 * `DruidCoordinator`

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by cryptoe** [comment]
UT's are incoming :)

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskMaster.java [hunk 1]
```diff
@@ -236,8 +227,10 @@ public Optional<String> getRedirectLocation()
     String leader = overlordLeaderSelector.getCurrentLeader();
     // do not redirect when
     // leader is not elected
-    // leader is the current node
-    if (leader == null || leader.isEmpty() || overlordLeaderSelector.isLeader()) {
+    // leader is the current node but not intialized.
+    if (leader == null
+        || leader.isEmpty()
+        || overlordLeaderSelector.isLeader() == DruidLeaderSelector.LeaderState.ELECTED) {
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskMaster.java:233]: Shouldn't this also include the case when `isLeader` returns `INITIALIZED`?

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskMaster.java [hunk 2]
```diff
@@ -170,7 +161,7 @@ public void stopBeingLeader()
       {
         giant.lock();
         try {
-          initialized = false;
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskMaster.java:173]: Doesn't removing this mean that we miss out on the state of initialization of this class itself?

File: server/src/main/java/org/apache/druid/rpc/ServiceClient.java [hunk 3]
```diff
@@ -37,7 +37,7 @@
  */
 public interface ServiceClient
 {
-  long MAX_REDIRECTS = 3;
+  long MAX_REDIRECTS = 5;
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/rpc/ServiceClient.java:40]: Maybe add a comment on why we need this value at 5.

File: server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java [hunk 4]
```diff
@@ -873,10 +901,14 @@ public void run()
...
         synchronized (lock) {
-          if (!coordLeaderSelector.isLeader()) {
-            log.info("LEGGO MY EGGO. [%s] is leader.", coordLeaderSelector.getCurrentLeader());
+          DruidLeaderSelector.LeaderState leaderState = coordLeaderSelector.isLeader();
+          if (leaderState == DruidLeaderSelector.LeaderState.NOT_ELECTED) {
+            log.info("I am no longer the leader. Current leader is [%s].", coordLeaderSelector.getCurrentLeader());
             stopBeingLeader();
             return;
+          } else if (leaderState == DruidLeaderSelector.LeaderState.ELECTED) {
+            throw new ISE(
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java:910]: Better to just log an error here and return.
An exception thrown here is just caught, logged and swallowed in `ScheduledExecutors.scheduledAtFixedRate()`.

File: server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java [hunk 5]
```diff
@@ -731,7 +759,7 @@ private void stopBeingLeader()
       lookupCoordinatorManager.stop();
       metadataRuleManager.stop();
       segmentsMetadataManager.stopPollingDatabasePeriodically();
-
+      exec.shutdownNow();
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java:762]: This need not be shutdown here as each `DutiesRunnable` which runs on this executor individually checks whether it should REPEAT or STOP.
If we do want to shutdown to be sure, however, then we must also re-initialize it in `becomeLeader`.

**Review by kfaraz — COMMENTED** [review]
Thanks for the changes, @cryptoe !
I have added some comments.

I also think there is probably a slightly simpler approach that might be more effective.
1. The `isLeader` method can remain as it is. Almost every consumer of the `isLeader` method wants a leader that is elected and fully initialized. So it can continue to return a boolean rather than the `LeaderState`. 
(We can rename it if we want to communicate that it is fully initialized).
2. It would be upto the actual impls (say `CuratorDruidLeaderSelector` and `K8sLeaderSelector`) to include the check on initialization in their `isLeader` methods.
3. Similary, the `DruidCoordinator` should get the invocation to `becomeLeader` only when things are fully initialized, rather than getting the trigger prematurely and having to keep checking on the state.
4. The `LeaderState` enum may or may not exist in the `DruidLeaderSelector` but its definitely not something that the consumers of the leader selector should have to concern themselves with.
5. The only real place this seems to matter is when determining redirection. For this, we should just add a separate method `LeaderSelector.isElected` rather than having to change the conditions everywhere else.

Let me know what you think.

**Comment by churromorales** [comment]
does this not affect the `druid-kubernetes-extensions` where it is k8s based?

**Pull request closed without merging**